### ISSUE TITLE
mincpmv4.5 training stuck

### ISSUE THREAD

**Issue by anxu829** [modelscope/ms-swift#5733]
**Describe the bug**
What the bug is, and how to reproduce, better with screenshots(描述bug以及复现过程，最好有截图)

my training data has following format : 
```
{
   'user' : '<image>describe following image',
   'assistant': ...,
   'user' : '<image>pic1 is a crop in pic2 , please return the bbox in pic2',
   'assistant': '<box> x , y , x , y </box>',
   'user' : 'is pic1 contain sth ?',
   'assistant' : 'yes/no'
}
```

i use following training config 

```
{
  "model_type" : "minicpmv4_5",
  "model": "/root/tmp/minicpm_v_4_5",
  "train_type" : "lora",
  "target_modules" : ["all-linear"],
  "lora_rank" : 16,
  "lora_alpha" : 64,
  "freeze_llm" : false,
  "freeze_vit": false,
  "freeze_aligner": false,
  "dataset": [
        ...
  ],
  "num_train_epochs" : 8,
  "max_length" : 3200,
  "max_pixels": 802816,
  "output_dir" : "~/0908",
  "per_device_train_batch_size" : 4 ,
  "gradient_accumulation_steps" : 2 ,
  "dataloader_num_workers": 4,
  "learning_rate" : 1e-5,
  "gradient_checkpointing" : true ,
  "max_grad_norm" : 1.0,
  "logging_steps" : 10,
  "warmup_ratio" : 0.03,
  "attn_impl": "flash_attn",
  "torch_dtype": "bfloat16",
  "save_strategy" : "steps",
  "save_steps": 420,
  "eval_strategy" : "steps",
  "eval_steps": 60,
  "gradient_checkpointing_kwargs" : {"use_reentrant": false}
}
```
and i use following start scripts:

(sft.py is just a copy sft.py in swift, change nothing)
```
torchrun --master_port 29501 --nproc-per-node 4  sft.py xxx.json
```

the traning will stuck like this : 

<img width="1528" height="373" alt="Image" src="https://github.com/user-attachments/assets/6e796e52-c8e1-4686-ad49-54d57a5cf567" />

i add some print in  MINICPMV::get_vllm_embedding:

```

        image_num = []
        for i in range(bs):
            cur_vs_hs = vision_hidden_states[i]
            cur_vllm_emb = vllm_embedding[i]

            if len(cur_vs_hs) == 0:
                updated_vllm_embedding[i] = cur_vllm_emb
                continue
                
            cur_image_bound = data['image_bound'][i]
            image_num.append(len(cur_image_bound))
            ... 
        print(image_num , vllm_embedding.shape )
```

also , i found a interesting thing : 

if i fix the inputs size in  template/minicpm.py::MiniCPMV4_5Template::_encode like this : 

```
    def _encode(self, inputs: StdTemplateInputs) -> Dict[str, Any]:
        encoded = Template._encode(self, inputs)
        images = inputs.images
        images = [im.resize((448,448 * N )) for im in images] # i try n = 1  or 10 , both will work
        ...

```
the training can continue smoothly

**Your hardware and system info**
Write your system info like CUDA version/system/GPU/torch version here(在这里给出硬件信息和系统信息，如CUDA版本，系统，GPU型号和torch版本等)

**Additional context**
Add any other context about the problem here(在这里补充其他信息)

**Comment by hjh0119** [comment]
I can’t reproduce this issue with the latest Swift. Could you retry after upgrading Swift?

By the way, it might be an OOM caused by very large images — please monitor GPU memory usage (e.g. with nvidia-smi) while reproducing.

**Comment by anxu829** [comment]
thank you for your trying to reproduce it, i tried the latest swift and i still get stuck ..but when i try to sample a subset (only has 3 sample from my dataset) for you to debug i found the program can train successfully , so i guess its a problem only can reproduce on some strange data . let me check my training data firstly : )
---- Replied Message ----
From ***@***.***> Date 09/10/2025 11:45 To ***@***.***> Cc ***@***.***>***@***.***> Subject Re: [modelscope/ms-swift] mincpmv4.5 training stuck (Issue #5733) 
hjh0119 left a comment (modelscope/ms-swift#5733)
I can’t reproduce this issue with the latest Swift. Could you retry after upgrading Swift?
By the way, it might be an OOM caused by very large images — please monitor GPU memory usage (e.g. with nvidia-smi) while reproducing.

**Comment by anxu829** [comment]
although i don't know why , i found a solution to this bug . 

my input has many image with shortest side less than 448 , some very tiny image linke 127 x 220 , i found when traning with this image , the grogram may stuck , i guesss some error  in the preprocessing of image  may caused by this small image , after i hack a padding function in my program : 

```
def pad_to_at_least_448(image: Image.Image) -> Image.Image:
    """
    不缩放、不裁剪。
    如果某一边 < 448，居中补白到 448；
    如果某一边 >= 448，保留全部内容。
    背景为白色。
    """
    original_width, original_height = image.size

    # 计算输出尺寸：每边至少 448
    new_width = max(original_width, 448)
    new_height = max(original_height, 448)

    # 创建白色背景
    new_image = Image.new("RGB", (new_width, new_height), (255, 255, 255))

    # 计算居中粘贴位置
    left = (new_width - original_width) // 2
    top = (new_height - original_height) // 2

    # 粘贴原图
    new_image.paste(image, (left, top))

    return new_image
```

the grogram will not get stucked  : )
